File: ios/ExpoOpusModule.swift
import ExpoModulesCore

public class ExpoOpusModule: Module {
  var opusDecoder: OpaquePointer? = nil
    
  public func definition() -> ModuleDefinition {
    Name("ExpoOpus")
      
    Function("opusStart") {
        var error: Int32 = 0
        self.opusDecoder = opus_decoder_create(16000, 1, &error)
    }
    
    Function("opusDecode") { (concatenatedPackets: Data, packetSize: Int) -> Data in
        var allDecodedPCM = Data()
        let packetCount = concatenatedPackets.count / packetSize
        
        for i in 0..<packetCount {
            let startIndex = i * packetSize
            let endIndex = min(startIndex + packetSize, concatenatedPackets.count)
            let packetData = concatenatedPackets.subdata(in: startIndex..<endIndex)
            
            var pcm = [Int16](repeating: 0, count: 5760)
            let decodedSamples = packetData.withUnsafeBytes { (bytes: UnsafeRawBufferPointer) in
                return Int(opus_decode(self.opusDecoder!, 
                                      bytes.bindMemory(to: UInt8.self).baseAddress, 
                                      opus_int32(packetData.count), 
                                      &pcm, 
                                      5760, 
                                      0))
            }
            
            if decodedSamples > 0 {
                // Convert Int16 samples to Data (raw bytes)
                let pcmData = Data(bytes: pcm, count: decodedSamples * MemoryLayout<Int16>.size)
                allDecodedPCM.append(pcmData)
            }
        }
        
        return allDecodedPCM
    }
    
    Function("opusStop") {
        opus_decoder_destroy(self.opusDecoder)
        self.opusDecoder = nil
    }
  }
}

File: android/src/main/java/expo/modules/opus/ExpoOpusModule.kt
package expo.modules.opus

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.ByteArrayOutputStream

class ExpoOpusModule : Module() {
  
  private val opus = OpusVedro()
  
  override fun definition() = ModuleDefinition {
    Name("ExpoOpus")

    Function("opusStart") {
      opus.decoderInit(16000, 1)
    }
    
    Function("opusDecode") { concatenatedPackets: ByteArray, packetSize: Int ->
      val allDecodedPCM = ByteArrayOutputStream()
      val packetCount = concatenatedPackets.size / packetSize
      
      for (i in 0 until packetCount) {
        val startIndex = i * packetSize
        val endIndex = minOf(startIndex + packetSize, concatenatedPackets.size)
        val packet = concatenatedPackets.copyOfRange(startIndex, endIndex)
        
        val pcmBuffer = ByteArray(5760 * 2)
        val decodedSamples = opus.decode(packet, pcmBuffer)
        
        if (decodedSamples > 0) {
          allDecodedPCM.write(pcmBuffer, 0, decodedSamples * 2)
        }
      }
      
      allDecodedPCM.toByteArray()
    }
    
    Function("opusDecodeSingle") { value: ByteArray ->
      val output = ByteArray(5760 * 2)
      val res = opus.decode(value, output)
      output.copyOfRange(0, res * 2)
    }
    
    Function("opusStop") {
      opus.decoderRelease()
    }
  }
}